components/web_discovery/browser/regex_util.h

```diff
@@ -0,0 +1,51 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_WEB_DISCOVERY_BROWSER_REGEX_UTIL_H_
+#define BRAVE_COMPONENTS_WEB_DISCOVERY_BROWSER_REGEX_UTIL_H_
+
+#include <deque>
+#include <memory>
+#include <optional>
+#include <string>
+
+#include "base/containers/flat_map.h"
+#include "base/no_destructor.h"
+#include "third_party/re2/src/re2/re2.h"
+
+namespace web_discovery {
+
+// Lazily creates and caches pre-compiled regexes, mainly used for
+// privacy risk assessment of page URLs/contents.
+// This class is not thread safe.
+class RegexUtil {
+ public:
+  RegexUtil();
+  ~RegexUtil();
+
+  RegexUtil(const RegexUtil&) = delete;
+  RegexUtil& operator=(const RegexUtil&) = delete;
+
+  bool CheckForEmail(std::string_view str);
+  bool CheckForLongNumber(std::string_view str, size_t max_length);
+  bool CheckPathAndQueryStringKeywords(std::string_view path_and_query);
+  bool CheckQueryStringOrRefKeywords(std::string_view str);
+  bool CheckQueryHTTPCredentials(std::string_view str);
+
+ private:
+  std::optional<re2::RE2> email_regex_;
+  // key is long number map length
+  base::flat_map<size_t, std::unique_ptr<re2::RE2>> long_number_regexes_;
+  std::deque<re2::RE2> path_and_query_string_keyword_regexes_;
+  std::deque<re2::RE2> query_string_and_ref_keyword_regexes_;
+  std::optional<re2::RE2> http_password_regex_;
+  std::optional<re2::RE2> non_alphanumeric_regex_;
+};
+
+static base::NoDestructor<RegexUtil> g_regex_util;
```

Comment:
pls check the proper way, it can't be in the header
https://source.chromium.org/chromium/chromium/src/+/main:base/no_destructor.h;l=52?q=no_destructor&ss=chromium

Reply:
nothing has changed?

Reply:
ah, looking at the wrong commit